오픈 마켓 II [STEP1] 아샌, 제인

ICS-Asan opened the pull request:
안녕하세요. 쿠마! @leejun6694

아샌(@ICS-Asan), 제인(@siwonkim0) 입니다. 

일정에 맞추느라 부족한 부분이 많겠지만 STEP1 리뷰 잘 부탁드립니다:)

작은 부분이라도 피드백 주시면 같이 고민하면서 피드백을 반영하여 코드를 개선할 수 있도록 하겠습니다!!

감사합니다!!

## 💻  주요 구현 내용

### < NewProduct>

- 인스턴스를 JSON타입으로 변환 할 수 있도록 Codable 프로토콜 채택
- API문서에 기본값이 있는 discountedPrice, stock은 이니셜라이저 기본값 사용
    - 입력이 없으면 기본값으로 인스턴스를 생성할 수 있도록 구현

### < NewProductImage>

- Image를 multipart-form으로 변환시 편의성을 위해 해당 타입 생성
- ImageData를 통해 인스턴스를 생성하고 활용 가능

### < AddProductViewController>

- ImagePicker
    - UIImagePickerController()를 사용하여 이미지를 사진앨범이나 카메라에서 가져오도록 구현
    - UIAlertController()을 통해 actionSheet 얼럿을 띄워 선택지를 제공함
    - UIImagePickerControllerDelegate, UINavigationControllerDelegate 사용
    - 이미지를 수정하여 추가할수 있도록 구현 : imagePickerController(_ : , didFinishPickingMediaWithInfo:) 메서드를 통해 이미지에 대한 정보를 가져옴
    - 이미지의 개수가 최대 5개가 넘지 않도록 5개가 되면 이미지추가버튼이 사라지도록 구현
    - 추가한 이미지를 클릭시 추가된 이미지 대신 새 이미지가 선택될수 있고, 이미지 추가버튼을 선택시 새 이미지를 추가할 수 있도록 구현
        - 모든 이미지를 버튼으로 구현하여 각각 tag 번호를 부여하여 각각의 이미지를 식별함
    
- textView Placeholder / 글자수제한
    - UITextViewDelegate를 사용하여 textField의 placeholder같은 기능 구현
    - textView(_ :, shouldChangeTextIn:, replacementText:) 메서드를 통해 글자수 제한 기능 구현

## 🤔 고민했던 부분

### < StackView에 사진을 추가하는 방법: UIStackView_insertArrangedSubView >

> 새로 추가되는 이미지를 넣는 자리에 대한 고민을 하였습니다. 
이미지추가버튼 뒤에 이미지가 추가되도록하려면 addArrangedSubView를 사용하면 되는데, 사용자의 입장에서 이미지추가버튼 앞에 새로운 이미지가 추가되는 것이 더 좋을 것 같다고 판단했습니다.
그래서 저희는 원하는 인덱스에 추가할 수 있는 insertArrangedSubView를 사용하였습니다.
> 

### < 제품 이미지 수정 기능: UIView_tag >

> 현재 저희는 사진을 추가하는 부분을 StackView로 구현을 하였습니다.
그리고 이미지를 눌렸을 때 수정하도록 기능을 구현해놓았는데 코드로는 어떤 이미지인지 파악을 하기가 힘들었습니다.
그래서 이미지에 tag를 설정하여 눌려진 이미지의 tag가 몇번인지 확인을 하고 해당 이미지를 바꿀 수 있도록 구현을 하였습니다.
> 

## ❓   질문드리고 싶은 내용

### < 뷰에서 입력받은 데이터 처리 >

> 현재 상품등록화면에서 사용자가 입력한 값을 데이터로 변환하여 서버에 POST하도록 구현해야 했습니다.뷰의 요소들을 모두 옵셔널 값인데 각 뷰의 값을 바인딩하고 입력값을 검증하려고 하니 반복되는 코드가 많아지고 데이터를 다루는 것이 복잡해지는 것 같았습니다.
이렇게 많은 뷰 요소들에서 값을 받아 데이터 처리를 해야할 때 어떤 방법으로 하면 간결하고 간단하게 처리할 수 있을지 궁금합니다.
> 

## 💡   조언 받고 싶은 내용

### < ViewController의 코드 정리 >

> 상품 등록 화면에 세세한 기능들이 많이 들어가 뷰 컨트롤러의 부피가 많이 커졌다고 생각이 되었습니다.
UIImagePickerController, Alert, TextViewPlaceHolder 등의 구현이 되어있는데 뷰의 요소와 연관이 되어있어 해당 ViewController에 구현을 해놓은 상태입니다.
쿠마의 경우에는 ViewController의 코드가 길어진 경우에 어떤식으로 분리를 하는지 저희는 어떻게 개선하면 좋을지 조언을 얻고 싶습니다.
> 

## ⚠️ 아직 구현하지 못한 부분

- 텍스트 입력시 키보드가 입력부분을 가리지 않도록 개선하는 것
- 사용자에게 받아야하는 값의 타입에 따라 키보드 속성을 변경하는 것
- 사용자 입력값의 유효성을 검사하여 어떤 부분의 입력이 잘못되었는지 확인하는 부분
- 접근제어자 설정

감사합니다👍

ICS-Asan committed:
미확인 파일 삭제

siwonkim0 committed:
feat: Currency, NewProduct 타입 생성 #10

- 상품등록시 사용할 상품 모델 NewProduct 타입 생성
- NewProduct와 ProductDetail에서 사용되는 currency 프로퍼티의 타입을 String -> Currency로 변경

ICS-Asan committed:
feat: NewProductImage 생성 및 구현 #10

- 상품등록시 사용할 Image의 타입으로 NewProductImage 구현

siwonkim0 committed:
feat: JSONParser에 encodeToDataString 메서드 생성 #10

- 상품등록시 상품 정보를 JSON 타입으로 변경하고, JSON 타입을 String 타입으로 변경하는 기능

ICS-Asan committed:
feat: createParams, createDataBody 메서드 구현 #10

- createDataBody에서 사용할 append메서드 Data extension 내부에 구현

siwonkim0 committed:
feat: addProduct, createPostURLRequest, createRequestBody 구현 #10

- multipart/form-data를 통한 상품등록 기능 구현
₩

ICS-Asan committed:
feat: 상품등록 화면 생성 및 화면 전환 구현 #11

siwonkim0 committed:
feat: 상품등록 화면 틀 구현 #11

- 세로 스크롤뷰를 활용하여 전체적인 화면 구현
- 가로 스크롤뷰와 스택뷰를 활용하여 이미지 선택부분 구현

ICS-Asan committed:
feat: textView 세팅 구현 #11

- textView 테두리 및 placeHolder 구현
- textView 입력 글자 수 제한(1000자)

siwonkim0 committed:
feat: UIImagePickerController 생성 및 Alert 구현 #11

ICS-Asan committed:
feat: 선택된 이미지 View 및 프로퍼티에 등록 기능 구현 #11

- addToStackView(), addToProductImages()
- alert구현 부분 함수분리

siwonkim0 committed:
feat: UIImagePickerController의 이미지 크롭기능 구현 #11

siwonkim0 committed:
feat: 추가된 기존 이미지 수정기능 구현 #11

- 이미지를 스택뷰에 추가할때 버튼으로 만들어서 추가
- 이미지 버튼의 태그를 활용하여 선택된 이미지의 인덱스를 확인

ICS-Asan committed:
feat: 제품 이미지 업로드 5개 제한 기능 구현 #11

- 이미지가 5개가 되면 추가버튼 숨김
- 버튼 숨겨질 때 AutoLayout이 깨지는 것을 방지하기 위해 버튼에 제약조건 추

siwonkim0 committed:
feat: 스택뷰 안의 버튼에서 이미지를 가져오는 로직 구현 #11

- 버튼의 tag를 활용함

ICS-Asan committed:
feat: Done버튼 탭하면 Post되는 기능 구현 #11

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+import UIKit
+
+class AddProductViewController: UIViewController {
+    
```
MARK 를 사용해서 property 와 life cycle method 등을 구분해준다면, 가독성이 올라갈 수 있습니다 : )

ICS-Asan replied:
AddProductViewController에 MARK주석의 사용과 extenstion을 사용하여 구분을 해주었습니다!
구분되어 조금 더 보기 편해진것 같습니다. 감사합니다!!
다른 파일들의 경우에는 다음 스텝을 구현하면서 추가할 수 있도록 하겠습니다.😀

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        super.viewDidLoad()
+        imagePicker.delegate = self
+        imagePicker.allowsEditing = true
+        self.navigationController?.navigationBar.topItem?.title = "상품등록"
```
self 가 명시되는 곳이 있고, 아닌 곳들이 있네요. 하나의 스타일로 정리해볼까요?ㅎㅎ

ICS-Asan replied:
self를 사용하지 않는 것으로 통일 하였습니다!!!

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        setupDescriptionTextView()
+    }
+    
+    @IBAction func tapCancelButton(_ sender: UIBarButtonItem) {
```
기타 method 들도 구분해서 extension 으로 구분한다면 어떨까요?

ICS-Asan replied:
`MARK: - IBAction Method`를 사용하여 구분해 주었습니다!

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        let lastTagNumber = productImageStackView.subviews.count - 1
+        guard lastTagNumber >= 1 else { return }
+        
+        for buttonTag in 1...productImageStackView.subviews.count - 1 {
```
이미 정의된 값을 사용하면 어떨까요?

ICS-Asan replied:
lastTagNumber를 사용하여 for문을 반복하도록 수정하였습니다😄

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+    func getProductImageFromButton(with tag: Int) {
+        let imageButton = view.viewWithTag(tag) as? UIButton
+        let image = imageButton?.imageView?.image
+        guard let imageData = image?.jpegData(compressionQuality: 0.1) else { return }
```
86~88번째 줄을 guard 로 한번에 묶는건 어떨까요?
`imageButton` 부터 nil 이라면, 의미 없이 다음 코드들이 진행되어버리네요.

ICS-Asan replied:
모든 옵셔널 값을 바인딩 처리해서 nil이 나오는 순간 메서드를 탈출할 수 있도록 개선하였습니다.👍

leejun6694 commented on OpenMarket/OpenMarket/Model/JSONModel/Currency.swift:
```diff
@@ -0,0 +1,14 @@
+import Foundation
+
+enum Currency: String, Codable {
+    case KRW
```
대문자...? 인 이유를 알 수 있을까요?

ICS-Asan replied:
Enum이 JSON에서 사용될 때는 rawValue를 통해 case별로 분기하여 데이터를 인코딩 디코딩 하는 것으로 알고 있습니다.
저희가 사용하는 Currency의 경우에는 화폐 단위라 "KRW", "USD"가 의미전달이 명확하고 대체할 용어가 생각나지 않았습니다.
그리고 JSON, KRW, USD 등 축약형이 대표적인 경우에는 축약형임을 표현하기 위해 대문자 그대로 작성 하는 것으로 알고 있었습니다.
이런 경우에도 case를 LowCamelCase로 작성하기 위해 풀어서("korean", "unitedStates") 작성하는 것이 옳은 방법일지 궁금합니다.

leejun6694 replied:
축약형이 대표적인 경우임에도 enum 의 case 가 대문자인 경우는 굉장히 생소하네요..ㅎㅎ (혹시 예시가 있을까요?)
제가 생각했던 건, `krw`, `usd` 이었습니다..!

https://docs.swift.org/swift-book/LanguageGuide/Enumerations.html
음.. 축약형을 살려야한다면 qrCode 도 QRCode 가 되어야하지 않을까 싶네요 @.@

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+            currency = Currency.KRW
+        case 1:
+            currency = Currency.USD
+        default:
```
`index` 값으로 `case` 가 구분되고 있어서, 추후 추가 구현이 있거나 타인이 수정을 하게 되는 경우 이해에 시간이 걸리거나 잘못 구현될 여지가 높은 코드로 보여집니다.
`Currency` 에 값을 하나 추가하거나 다른 방법으로, `switch` 를 사용하지 않거나, 케이스를 좀 더 안전하게 나누는 방법을 고민해보면 좋을 것 같네요.

ICS-Asan replied:
조언 감사합니다!!
저희가 고민해봤는데 현재 SegementedControl의 title을 이용하여 구분하면 index와 관계없이 조금 더 명확하게 사용 할 수 있을 것 같다고 생각해서 다음과 같이 수정해봤습니다.
```swift
let selectedIndex = currencySegmentedControl.selectedSegmentIndex
guard let currentTitle = currencySegmentedControl.titleForSegment(at: selectedIndex) else { return }
if currentTitle == Currency.KRW.unit {
    currency = Currency.KRW
} else if currentTitle == Currency.USD.unit {
    currency = Currency.USD
}
```

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        }
+        addProductImage(with: image)
+        if productImageStackView.subviews.count == 6 {
+            self.addImageButton.isHidden = true
```
여기도... self

ICS-Asan replied:
이 부분도 self를 사용하지 않는 것으로 통일 하였습니다!!!

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+
+extension AddProductViewController: UITextViewDelegate {
+    
+    func setTextViewOutLine() {
```
`delegate` 관련 method 라기보단, `UI` 와 관련되어 보이네요 : )

ICS-Asan replied:
TextView를 Setup하는 부분과 Delegate 관련 메스드를 extenstion으로 구분해 주었습니다😀

leejun6694 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -46,6 +50,63 @@ class APIManager {
+    }
+    
+    func createMultiPartFormData(withParameters params: Parameters?, images: [NewProductImage]?) -> Data {
+        
```
불필요한 줄바꿈이 있네요.

ICS-Asan replied:
줄바꿈 수정했습니다 감사합니다^^

leejun6694 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -46,6 +50,63 @@ class APIManager {
 }
 
+extension APIManager {
+    func addProduct(product: NewProduct, images: [NewProductImage], completion: @escaping (Result<ProductDetail, Error>) -> Void) {
```
호출되는 곳에선 `Product(product: ~~)` 요렇게 되겠네요ㅠ

leejun6694 replied:
다른 API 가 필요하다면, APIManager 코드가 점점 길어지겠네요ㅠ
`API method` 와 `network method` 를 나눠보는 것도 역할이 나눠지는 방법 중 하나가 될 것 같습니다.

siwonkim0 replied:
API method와 network method를 나눠보라는 말씀이 혹시 둘중 하나를 extension으로 빼서 파일을 분리하라는 말씀이신가요..? 아니면 따로 타입을 만들어서 분리를 하는 것이 좋을까요,,? 저희가 코멘트를 제대로 이해하지 못한것같습니다.. ㅜㅜ 조금만 더 자세히 설명해주실수 있을까요....? 

일단 저희가 생각해본바로는 createRequestBody, createParams, createMultiPartFormData 이 세 메서드는 네트워크 통신과는 직접적인 관련이 없는 body를 만드는 메서드들이기때문에 따로 분리를 할 수 있다고 말씀하신게 아닐까 라고 생각해보았습니다..! 
저희가 이해한바가 맞을까요? 🧐

leejun6694 replied:
음.. 비슷하지만 조금 다른 부분이 있습니다ㅎㅎ
디스코드로 연락드릴게요~

leejun6694 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -46,6 +50,63 @@ class APIManager {
+        return params
+    }
+    
+    func createMultiPartFormData(withParameters params: Parameters?, images: [NewProductImage]?) -> Data {
```
parameter 가 어떤 곳에선 `with` 만 있고, 어떤 곳에선 `with+parameter` 가 되어 있네요
스타일을 하나로 맞추면 더 좋을 것 같습니다.

siwonkim0 replied:
parameter 삭제해서 스타일 맞춰주었습니다 ~~! 감사합니다

leejun6694 reviewed (changes requested):
아샌, 제인 안녕하세요!
step 1 구현으로 고생 많으셨습니다ㅎㅎ
몇가지 리뷰를 남겼는데, 확인 부탁드려요.
여러군데 해당하는 리뷰를 한군데만 남긴 것도 있으니, 이 점도 참고해주시면 감사합니다~~!

+
구조적인 부분이나, 생각이 필요한 부분도 체크를 해두었습니다.
당장 이번 스텝에서 구현이 어려울 수 있습니다...!
너무 욕심내지는 마시고, 가능하다면 생각+구현을, 시간이 오래걸릴 것 같고 감이 아직 안온다면 추후에 천천히 확인 부탁드려요~!

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+    @IBOutlet weak var productImageStackView: UIStackView!
+    @IBOutlet weak var addImageButton: UIButton!
+    
+    let apiManager = APIManager()
```
API 호출 매니저의 경우, 아래와 같이 호출할 수 있겠네요

- 현재처럼 `인스턴스` 생성 후 호출
- 매니저를 `singleton` 으로 구현
- 호출 method 를 `static` 으로 정의

어떤 장단점들이 있는지 생각해본다면 좋은 경험이 될 것 같습니다~~!

leejun6694 replied:
바로 사용하지 않는 property 라면 `lazy` 등을 활용할 수 있습니다~!

siwonkim0 replied:
APIManager 인스턴스가 현재 OpenMarketViewController, AddProductViewController 두군데서 사용되고 있습니다. 사용시마다 인스턴스를 만들어주는것보다 싱글턴으로 만들어서 불러오는 것이 더 효율적일 것 같아 싱글턴으로 변경하였습니다. 

다만 test 코드에서 Mock URLSession을 사용하고 있는데 이때 이니셜라이저를 통해 URLSession을 바꿔주고있어서 private init 으로 만들어주기는 어려울 것 같습니다.

이러한 경우에 그냥 타입 프로퍼티로 APIManager()을 들고있고, 이니셜라이저는 internal 그대로 사용해도 되는지 궁금합니다.

test를 위해서 이렇게 코드를 불완전하게(?) 구현하는것이 맞는지 아니면 test를 깔끔하게 포기하고 코드의 안정성을 높이는 것이 더 나은지 쿠마의 의견이 궁금합니다...!

```Swift
class APIManager {
  static let shared = APIManager()
  let urlSession: URLSessionProtocol

  init(urlSession: URLSessionProtocol = URLSession.shared) {
      self.urlSession = urlSession
  }
}
```

leejun6694 commented on OpenMarket/OpenMarket/Model/APIManager.swift:
```diff
@@ -1,7 +1,11 @@
 import Foundation
+import UIKit
+
+typealias Parameters = [String: String]
```
👍👍

leejun6694 commented on OpenMarket/OpenMarket/Controller/AddProductViewController.swift:
```diff
@@ -0,0 +1,218 @@
+        button.heightAnchor.constraint(equalTo: productImageStackView.heightAnchor).isActive = true
+        button.widthAnchor.constraint(equalTo: button.heightAnchor).isActive = true
+    }
+    
```
줄바꿈..!

ICS-Asan replied:
줄바꿈 수정했습니다 감사합니다😅

leejun6694 commented:
> 뷰에서 입력받은 데이터 처리

음.. 결국 어디선가는 가져와서 처리를 해야하는 값들이여서, 좋은 방법이 바로 생각나진 않네요.
TextField 등을 커스터마이징해서, `.text` 나 `type` 변환 등을 숨기고 명확하게 할 순 있지만, 아쉬운 점도 없지 않아 있는 것 같습니다.
혹시 다른 캠퍼분들 중 좋은 아이디어가 있다면, 저도 공유 부탁드립니다...ㅎㅎ

> ViewController의 코드 정리

MVC.. `Massive View Controller`..... iOS 개발자라면 언젠가 부딪히는 고민이라고 생각됩니다.
`Delegate` 관련 코드들은 당장 어쩔 수 없다고 하더라도, `Alert` 는 별개로 구현이 가능할 것 같네요.
허나, 이 고민은 당장 이번 스텝에서 모든걸 해결할 순 없는 문제라고 생각됩니다. 이런저런 포인트에서 과연 이 구현을 **Controller** 에서 해야하는가? 라는 생각을 남겨두기만 하더라도, 좋은 겸험이 아닐까 생각되네요.
필요한 고민이지만, 이번 스텝에선 다른 부분들에 더 시간을 쏟으면 어떨까 의견드립니다...!

siwonkim0 committed:
chore: 코딩 컨벤션 수정 #12

- MARK 주석 사용하여 메서드 분리
- 줄바꿈 수정
- self 키워드 사용 통일
- identifier, secret 아샌걸로 변경

ICS-Asan committed:
refactor: 불필요한 로직 수정 #12

- addToProductImages 내부 연산된 값을 사용하여 같은 과정이 반복되는 부분 삭제
- getProductImageFromButton 내부 모든 상수 옵셔널 바인딩 처리
- createNewProduct 내부 segmentedControl의 title을 이용하여 Currency 구분

siwonkim0 committed:
refactor: APIManager 싱글턴 패턴 적용 #12

ICS-Asan committed:
refactor: parameter, NewProduct 네이밍 수정 #12

- parameter가 사용되는 부분 통일성을 위해 with params로 수정
- 명확한 의미전달을 위해 NewProduct->NewProductInformation 수정

siwonkim0 committed:
refactor: createNewProduct 메서드 전면수정 #12

- post 할때 입력받은 입력값들을 검증하는 로직 수정

ICS-Asan committed:
feat: 키보드 관련 기능 구현 #12

- 키보드의 높이만큼 scroll View의 높이가 줄어들도록 구현
- 빈 공간을 터치했을 때 키보드가 내려가도록 구현
- TextField의 입력값에 맞는 키보드가 나타나도록 설정(Storyboard)

ICS-Asan commented:
세세한 피드백 감사합니다 쿠마!!
리팩토링과 미구현된 부분을 구현해보면서 많은 고민들을 하느라 시간도 오래 걸리고 코드도 많이 추가 되었습니다.
1. 뷰에서 입력받은 데이터 처리
  - 뷰에서 입력받은 데이터를 검증하여 담아두기 위한 enum을 생성하여 활용하는 방법으로 구현을 해보았습니다.
  - MARK: - Create Data To Post 부분에 구현을 해두었는데 현재 방법에 대한 쿠마의 의견이 궁금합니다.
  - 현재 단순 print로 확인하고 있는 부분은 Alert를 통해 메세지를 전달해줄 예정입니다.
2. 키보드 관련
 - 키보드의 높이에 맞춰 scrollView의 높이를 조절하는 방식으로 키보드의 가림 현상을 해결하였습니다.
 - storyboard에서 각 textField의 입력값에 맞는 키보드가 띄워지도록 설정하였습니다.
 - 빈 공간을 탭하면 키보드가 내려가도록 gesture를 추가해 주었습니다.

siwonkim0 committed:
feat: Alert 관련 기능 구현 #12

- 상품등록시 textField의 입력값에 따라 Alert을 띄우도록 구현
- JSONParser의 encodeToDataString 메서드에 keyEncodingStrategy 추가하여 할인금액이 post되지않는 문제 해결
